### ISSUE TITLE
Adding ColorPicker and Related Primitive Controls

### ISSUE THREAD

**Issue by robloo** [AvaloniaUI/Avalonia#4179]
# Adding ColorPicker/Button and ColorCanvas Controls

I don't think Avalonia has a ColorPicker control. Is there a plan to add this to the project? If so, I've been implementing my own (as many others have) and would like to propose the below idea. This is currently for UWP but porting should be possible. Right now I think it provides a unique concept to discuss and see if it fits with this project.

**ColorCanvas**

See WinUI

**ColorPicker Example**

Newest example:

![image](https://user-images.githubusercontent.com/17993847/162509229-d8230e1a-165f-4e56-a39d-c4e0e43238e2.png)

Older example from the Windows Community Toolkit

![example1](https://user-images.githubusercontent.com/17993847/87890662-69fc2180-ca05-11ea-83c5-3ea112f09f5e.gif)

| Spectrum    | Palette | Channels |
| ------------- | ------------- | ------------- |
| ![Spectrum](https://user-images.githubusercontent.com/17993847/87890675-72ecf300-ca05-11ea-8ae5-94fcfd80e2a1.png) | ![Palette](https://user-images.githubusercontent.com/17993847/87890688-7aac9780-ca05-11ea-9a3b-ce7ba60b620c.png) | ![Channels1](https://user-images.githubusercontent.com/17993847/87890763-cfe8a900-ca05-11ea-9659-4c112224e382.png) |
| | | ![Channels2](https://user-images.githubusercontent.com/17993847/87890701-85ffc300-ca05-11ea-8169-a1bc5505b23c.png) |

**Comment by maxkatz6** [comment]
I am trying to implement ColorPicker from WinUI in Avalonia.
I will provide some results in a few days.

By the way, I like how your ColorPicker looks!

**Comment by robloo** [comment]
Awesome, I'm glad someone is already working on this!

The main issues with the WinUI/UWP ColorPicker I find are poor styling/layout, large size that isn't adaptive, and it doesn't fit in with other pickers (no flyout). Since AvaloniaUI is fixing past mistakes whenever possible it might make sense to correct those here. Alternatively, we could go with a two control approach which is what I'm advocating for in UWP:

 1. ColorPicker : A panel that can be used to select a color like WinUI
 2. ColorSwatchPicker : An actual picker that previews a selected color and opens a flyout to edit (like DatePicker, TimePicker, etc.) This is far more user-friendly and is what I pictured above.

**Comment by robloo** [comment]
@grokys @danwalmsley Regardless of if this 'ColorSwatchPicker' gets into the windows community toolkit or WinUI would you consider it for this project? If so, I would be happy to provide a port as I'm already doing for the UWP repos.

As stated above, I propose two controls: (1) the ColorPicker similar to UWP that @maxkatz6 is already working on and (2) This flyout-style ColorSwatchPicker useful for constrained spaces or where form-style input is needed.

**Comment by robloo** [comment]
I got the go ahead to start adding the flyout-style ColorPickerButton to the UWP toolkit. I still think it would be very useful in Avalonia so would like some additional feedback on this issue.

It also became apparent that the UWP ColorPicker is incorrectly named. It is not a Picker like the other controls: DatePicker, TimePicker, etc. It would be better to correct this in Avalonia now and follow the WPF extended toolkit convention: https://www.codeproject.com/Articles/779105/Color-Canvas-and-Color-Picker-WPF-Toolkit

 1. ColorCanvas : A panel that can be used to select a color like WinUI. Matches what is currently in UWP.
 2. ColorPicker : The control I'm proposing to add. An actual picker that previews a selected color and opens a flyout to edit (like DatePicker, TimePicker, etc.) This is far more user-friendly and is what I pictured above.

**Comment by maxkatz6** [comment]
@robloo I have pushed my ColorPicker port from WinUI - https://github.com/maxkatz6/Avalonia.ColorPicker
It still isn't ready, although it mostly works. 
I haven't a lot of time to work on it, however I still want to complete it.

**Comment by robloo** [comment]
@maxkatz6 Good stuff! I'm starting my own port to the Uno Platform from the WinUI sources as well. I'm not going to use yours as a base so I don't have to worry about the Avalonia-specific changes and also so I can document the architecture in WinUI. That said, it will be good to do a diff of the two C# versions once complete too see if there are any mistakes.

**Comment by robloo** [comment]
@maxkatz6 The Uno Platform C# port of this was completed as well. I'll circle back to the ColorPickerButton in the Windows Community Toolkit by next week and hopefully bring that to completion. The Toolkit maintainer wants to completely re-style the ColorPicker itself to get some of these updates. However, we can ignore those for now and just bring in the Button in a month or two when its ready. It should be a lot easier as it's already C# and only depends on the ColorSpectrum.

**Comment by robloo** [comment]
@maxkatz6 Would be great to get this in for the next 0.11 release. I'll port over the ColorPickerButton once you have merged in at least the ColorSpectrum. There can likely then be some larger architectural changes to modernize things and make the control responsive (something the WinUI version lacks).

**Comment by pr8x** [comment]
Would be nice to integrate that into the DevTools as well to allow for easier styling etc.

**Comment by robloo** [comment]
With Flyout almost added that unblocks DropDownButton which I plan to have implemented along with the other remaining WinUI button types. With DropDownButton I think this ColorPicker concept itself becomes unblocked.

I still have not figured out exactly what the implementation will look like. I go back and forth between doing a clean reimplementation fixing all the wrongs with WinUI or just porting over directly from the WCT and using @maxkatz6 port as-is. The WPF toolkit had better names for these controls though (WinUI's ColorPicker is really a ColorCanvas and the ColorPickerButton is really what the ColorPicker should be). 

There are other architectural/naming changes I would like to make based on [this document](https://github.com/microsoft/microsoft-ui-xaml/blob/master/dev/ColorPicker/improvements.md) and discussed a little with @maxkatz6 [here](https://github.com/maxkatz6/AvaloniaWinUI.ColorPicker/issues/1#issuecomment-691764669). This would align the control quite a bit more with XAML concepts.

What I want to get a feel for is how people would accept larger deviations from WinUI such as the control names themselves and the property name updates (ThirdDimension -> ThirdChannel) etc. Any comments?

**Comment by robloo** [comment]
Here is the plan:

 1. Deviate from all previous implementations will be made so that this control fully integrates with Avalonia and fixes past mistakes
 2. The following controls will be added:
     * ColorPicker : This is the main control to use for color selection and **appears as a drop down button**. All other pickers are already drop downs so ColorPicker was originally named incorrectly in WinUI (this was discussed in that repo).
     * ColorSpecturm [Primitive] : This will be a refactored port of the WinUI code. Features are largely the same and perhaps only a few properties are renamed.
     * ColorPickerSlider [Primitive] : This is NOT based on the WinUI code and is a complete rewrite.
     * ColorPreviewer [Primitive] : A control that previews a color with optional accent colors to the side. This is brand new.
     * ColorEditor [Primitive] : This is the 'canvas' version of the control that does not have a drop down button. ColorPicker will host this in a drop down button flyout.
 3. Everything is going to go inside a new ColorPicker directory -- including the primitives

Each of these components will be implemented in a separate PR so there is a chance it can be reviewed sanely:

 1. ColorSpectrum
 2. ColorPickerSlider
 3. ColorPreviewer
 4. ColorEditor
 5. ColorPicker

There is an open question of naming. The WinUI code isn't even consistent here.

 * "ColorChannel" vs "ColorComponent". "Component" is more user-friendly and matches with iOS for example, however it is less technical than "Channel". `ColorSpectrum` in WinUI uses the term "Component" but the `ColorPickerSlider` uses "Channel".
 * "ColorRepresentation" vs "ColorModel" vs "ColorSpace". "ColorRepresentation" is what is used in WinUI but I'm changing this to "ColorModel" which is shorter and actually makes more sense. ColorSpace is not always the correct term.

**Comment by robloo** [comment]
Just going to leave this here:

![image](https://user-images.githubusercontent.com/17993847/162509229-d8230e1a-165f-4e56-a39d-c4e0e43238e2.png)

**Comment by maxkatz6** [comment]
@robloo may I ask, do you plan to port SwitchPresenter as well or use styled TabControl? TabStrip? From what I see, we don't really need a new control here.

**Comment by robloo** [comment]
> do you plan to port SwitchPresenter as well or use styled TabControl? TabStrip? From what I see, we don't really need a new control here.

No, I don't plan to port SwitchPresenter. I have the same conclusion that TabStrip is enough.
